src/ssr-template-helper.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Capability} from './service/viewer-impl';
+
+/**
+ * Helper, that manages the proxying of template rendering to the viewer.
+ */
+export class SsrTemplateHelper {
+
+  /**
+   * @param {string} sourceComponent
+   * @param {!./service/viewer-impl.Viewer} viewer
+   * @param {!./service/templates-impl.Templates} templates
+   */
+  constructor(sourceComponent, viewer, templates) {
+
+    /** @private */
+    this.viewer_ = viewer;
+
+    /** @private */
+    this.templates_ = templates;
+
+    /** @private {!XMLSerializer} */
+    this.xmls_ = new XMLSerializer();
+
+    this.sourceComponent = sourceComponent;
+
+    /** @const @private {boolean} */
+    this.viewerCanRenderTemplate_ = this.viewer_.canRenderTemplates();
+  }
+
+  /**
+   * Whether the viewer can render templates.
+   * @return {boolean}
+   */
+  isSupported() {
+    return this.viewerCanRenderTemplate_;
+  }
+
+  /**
+   * Proxies xhr and template rendering to the viewer and renders
+   * the response.
+   * @param {!Element} element
+   *     viewer to determine the component responsible for proxying the
+   *     request.
+   * @param {?function():Promise<?>} renderTemplateSuccessCallback
+   * @param {?function():Promise<?>} renderTemplateFailureCallback
+   * return {!Promise}
+   */
+  fetchAndRenderTemplate(
+    element,
+    renderTemplateSuccessCallback,
+    renderTemplateFailureCallback) {
+    const inputsAsJson = this.getElementInputsAsJson_(element);
+    const elementAttrsAsJson = this.getElementAttributesAsJson_(element);
+    elementAttrsAsJson.inputData = inputsAsJson;
+    const mustacheTemplate = this.xmls_.serializeToString(
+        this.templates_.findTemplate(element));
+    return this.viewer_.sendMessageAwaitResponse(
+        Capability.VIEWER_RENDER_TEMPLATE,
+        {
+          data: elementAttrsAsJson,
+          mustacheTemplate,
+          'sourceAmpComponent': this.sourceComponent,
+        })
+        .then(resp => {
+          if (renderTemplateSuccessCallback) {
```

Comment:
Are we going to somehow check the AmpCors Headers?
I remember Justin suggested once to strip out amp cors logic out of xhr service so that we could also share it here.

Reply:
ah yes, this should handled. I believe Will mentioned this before but then we became focus on how to abstract the code out of the extensions so that the implementation is cleaner. I'll make these changes..maybe it would even make sense to do it in a separate PR along with the typedef changes you mentioned...as stripping out the logic from xhr for sharing will introduce more changes in this PR and make it more difficult to manage. Thanks

Reply:
SG, thanks!

About the request and response type, I have a concern that if that's compatible with what we already have for xhr interceptor.
There are some XHRs that doesn't contain a template which will go through xhr-service logic and got intercepted by XHR intercepter (for example, each AMP document will fetch a service worker when it's loaded). Ideally, we could like to have a compatible interface for both template-helper and xhr-intercepter, so in the viewer we can treat them as one.

For example,
I believe the request type in xhr-intercepter is something like:
{
  'originalRequest': xhr requests
}
https://github.com/ampproject/amphtml/blob/master/src/service/xhr-impl.js#L204

If we could define the request type of template-helper to be 
{
  'originalRequest': xhr requests,
  `success_template`:
  `failure_template`:
  ....
}
Then this could be compatible for both.

The response type would be trickier because I know we didn't nested the response in xhr-intercepter..... Probably we should do that first.

Hope this makes sense, we can definitely discuss this more.

Reply:
Yes we can discuss this further offline...though I do have a basic understanding of what is desired. Thanks